tools/compression/src/compress.c

```diff
@@ -0,0 +1,283 @@
+	for (index = 0; index < sections - 1; index++) {
+		/* Read block into read_buf */
+		tptr = read_buf;
+		readsize = block_size;
+		while (readsize > 0) {
+
+			nbytes = read(in_fd, tptr, readsize);
+			if (nbytes < 0) {
+				/* EINTR just means that we received a signal */
+
+				if (nbytes != -EINTR) {
+					printf("Read failed: %d\n", (int)nbytes);
+					goto error;
+				}
+			} else if (nbytes == 0) {
+				printf("Unexpected end of file\n");
+				break;
+			} else {
+				readsize -= nbytes;
+				tptr += nbytes;
+			}
+		}
+		writesize = block_size;
+#if CONFIG_COMPRESSION_TYPE == 1
+		/* LZMA Compression for data in read_buf into out_buf */
+		ret = LzmaCompress(&out_buf[LZMA_PROPS_SIZE], &writesize, read_buf, (block_size - readsize), out_buf, &propsSize, -1, 0, -1, -1, -1, -1, -1);
+		if (ret != SZ_OK) {
+			printf("LZMA Compress failed, ret = %d\n", ret);
+		}
+
+		printf("==> lzma_compress %d writesize %lu\n", index, writesize);
+#else
+		printf("Compression for type %d not supported\n", CONFIG_COMPRESSION_TYPE);
+		printf("Set CONFIG_COMPRESSION_TYPE to %d, then generate mkcompressimg again for this type", CONFIG_COMPRESSION_TYPE);
+		exit(1);
+#endif
+		phdr->secoff[index + 1] = phdr->secoff[index] + writesize;
+
+		/* Write out_buf to output file */
+		tptr = out_buf;
+		while (writesize > 0) {
+			nbytes = write(out_fd, tptr, writesize);
+			if (nbytes < 0) {
+				/* EINTR just means that we received a signal */
+
+				if (nbytes != -EINTR) {
+					printf("Write failed: %d\n", (int)nbytes);
+					goto error;
+				}
+			} else if (nbytes == 0) {
+				printf("Unexpected end of file\n");
+				break;
+			} else {
+				writesize -= nbytes;
+				tptr += nbytes;
+			}
+		}
+	}
+
+	rpos = lseek(out_fd, 0, SEEK_SET);
+	if (rpos != 0) {
+		printf("Failed to seek to position %lu\n", (unsigned long)size);
+		goto error;
+	}
+
+	/* Write header into output file */
+	writesize = size;
+	tptr = (unsigned char *)phdr;
+	while (writesize > 0) {
+		nbytes = write(out_fd, tptr, writesize);
+		if (nbytes < 0) {
+			/* EINTR just means that we received a signal */
+
+			if (nbytes != -EINTR) {
+				printf("Write failed: %d\n", (int)nbytes);
+				goto error;
+			}
+		} else if (nbytes == 0) {
+			printf("Unexpected end of file\n");
+			break;
+		} else {
+			writesize -= nbytes;
+			tptr += nbytes;
+		}
+	}
+
+error:
+	if (phdr) {
+		free(phdr);
+	}
+	if (read_buf) {
+		free(read_buf);
+	}
+	if (out_buf) {
+		free(out_buf);
+	}
+	close(in_fd);
+	close(out_fd);
+	return;
+
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+int main(int argc, char *argv[])	//Main defined
+{
+	char *p;
+	int ret;
+	int block_size;
+	int comp_format;
+	struct stat buf;
+
+	if (argc != 5) {
+		fprintf(stderr, "Unexpected number of arguments\n");
+		show_usage(argv[0]);
+	}
+
+	block_size = strtol(argv[1], &p, 10);
```

Comment:
since block_size is of int type, better to use stoi.

Reply:
made this change.
changed to atoi